File: python2/sort/M1SelectionSort.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os, sys


class M1SelectionSort:

    # 外层控制循环轮数，外层控制循环次数，次数随轮数收敛
    # 每轮筛选出对应轮数的极值，往前排，，下轮循环起点对应后移
    # 时间复杂度 o(n^2) 双层循环
    # 空间复杂度 o(1) 只申请了一个额外空间
    # 不稳定排序：相同元素2,2经过排序后顺序会倒置，理论上稳定排序应该是平移关系
    @staticmethod
    def solve1(arr):
        print(arr)
        size = len(arr)
        min_pos = 0
        for i in range(size-1):
            for j in range(i,size):
                if arr[min_pos] > arr[j]:
                    min_pos = j
            M1SelectionSort.swap(arr,i,min_pos)
        print(arr)


    # 选择排序优化解(高低位排序)
    # 1.外层循环为轮数，内层循环为每轮次数
    # 2.每轮循环同时找最大值，最最小值，当前轮循环次数2倍速度收敛
    # 3.调换元素时，先调换者可能对后调换者有扰动，需要通过分支判断
    # 时间复杂度 o(n^2)
    # 空间复杂度 o(2)
    # 稳定性：排序过程相同元素相对顺序可能颠倒，不稳定
    # 计算时间复杂度时初始化语句和打印语句，开销不考虑，重点关注递归和循环，忽略常数项、低次项和系数
    @staticmethod
    def solve2(arr):
        print(arr)
        size = len(arr)
        loop = (size+1)/2
        for i in range(loop):
            min_pos = i
            max_pos = size-1-i
            if arr[max_pos]<arr[min_pos]:
                M1SelectionSort.swap(arr,min_pos,max_pos)
            for j in range(i+1,size-1-i):
                if arr[min_pos]>arr[j]:
                    min_pos = j
                if arr[max_pos]<arr[j]:
                    max_pos = j
            M1SelectionSort.swap(arr,i,min_pos)
            if max_pos == i:
                M1SelectionSort.swap(arr,size-1-i,min_pos)
            else:
                M1SelectionSort.swap(arr, size - 1 - i, max_pos)
        print(arr)

    @staticmethod
    def swap(arr,i,j):
        arr[i],arr[j] = arr[j],arr[i]

File: src/test/java/com/bigdata/sort/M1SelectionSort.java
package com.bigdata.sort;

import org.junit.Test;

/**
 * 选择排序优化版
 * 选择排序，不稳定，实际生产不用
 */
public class M1SelectionSort extends BaseSort{

    /**
     * 选择排序：
     * 1.外层控制循环轮数，内层控制每轮循环次数，内层循环次数，随外层轮数增大收敛;
     * 2.每轮循环筛选出对应轮数极值，放置到前面
     * 时间复杂度 o(n^2) 双层循环，最好，最坏时间复杂度都为o(n^2)
     * 空间复杂度 o(1) 只申请一个交换空间
     * 稳定性：相同元素在排序过程中可能颠倒顺序，所以是不稳定的
     * DEBUG - (5),8,5,(2),9 （相同元素5的相对位置发送了改变）
     * DEBUG - 2,(8),(5),5,9
     * DEBUG - 2,5,(8),(5),9
     * DEBUG - 2,5,5,(8),9
     */

    @Test
    public void testSolve(){
//        solve1(arr);
        solve2(arr);
    }

    public static void solve1(int[] arr){
        int size = arr.length;
        for(int i=0;i<size-1;i++){
            int minPos = i;
            for(int j=i+1;j<size;j++){
                if(arr[minPos]>arr[j]){
                    minPos = j;
                }
            }
            int temp = arr[i];
            arr[i] = arr[minPos];
            arr[minPos] = temp;
        }
    }

    /**
     * 选择排序优化解(高低位排序)
     * 1.外层循环为轮数，内层循环为每轮次数
     * 2.每轮循环同时找最大值，最最小值，当前轮循环次数2倍速度收敛
     * 3.调换元素时，先调换者可能对后调换者有扰动，需要通过分支判断
     * 时间复杂度 o(n^2)
     * 空间复杂度 o(2)
     * 稳定性：排序过程相同元素相对顺序可能颠倒，不稳定
     * 计算时间复杂度时初始化语句和打印语句，开销不考虑，重点关注递归和循环，忽略常数项、低次项和系数
     */
    public static void solve2(int[] arr){
        int length = arr.length;
        // 循环轮数 由于每轮循环同时找最大和最小值，因此轮数减半
        int loop = (length+1)/2;
        for(int i=0;i<loop;i++){
            // 初始化最大，最小下标
            int minPos = i;
            int maxPos = length-1-i;
            // 最大最小下标先比较
            if(arr[minPos]>arr[maxPos]){
                swap(arr,minPos,maxPos);
            }
            // 循环找最大值和最小值，两边往中间收敛
            for(int j=i+1;j<length-i;j++){
                if(arr[minPos]>arr[j]){
                    minPos = j;
                }
                if(arr[maxPos]<arr[j]){
                    maxPos = j;
                }
            }
            // 最小值往前放
            swap(arr,i,minPos);

            // 最大值不在刚刚被替换位置时，直接将最大值往后放，否则将最大值与上面调换过后位置对调（此时minPos 实质是i）
            if(i!=maxPos){
                swap(arr,length-1-i,maxPos);
            }else{
                swap(arr,length-1-i,minPos);
            }
        }
    }

}
